command/agent/agent.go

```diff
@@ -507,7 +507,14 @@ func convertServerConfig(agentConfig *Config) (*nomad.Config, error) {
 		conf.FailoverHeartbeatTTL = failoverTTL
 	}
 
-	if *agentConfig.Consul.AutoAdvertise && agentConfig.Consul.ServerServiceName == "" {
+	// Add the Consul and Vault configs
+	conf.ConsulConfigs = helper.SliceToMap[map[string]*config.ConsulConfig](
+		agentConfig.Consuls,
+		func(cfg *config.ConsulConfig) string { return cfg.Name },
+	)
+
+	consul := conf.ConsulConfigs[structs.ConsulDefaultCluster]
+	if *consul.AutoAdvertise && consul.ServerServiceName == "" {
 		return nil, fmt.Errorf("server_service_name must be set when auto_advertise is enabled")
 	}
```

Comment:
The non-default cluster will have these fields set because they're getting merged onto the `DefaultConsulConfig`. We don't do any validation until well after that merge, so it's possible to do this but kind of painful at the moment. I can open an issue to cover this in future refactoring though.

Reply:
Opened https://github.com/hashicorp/nomad/issues/19029